### PR TITLE
Cleanup fix

### LINKED ISSUE #64: Cleaning up the React Upload Widget instance on unload
hey :)

i use the react upload widget in my app. 
i found that if i navigate from the upload page(in my app it /addnew) to the home page /   and the back the /addnew
the iframe is still exist as expected . but when i try to updateUrl state something related to the iframe is going wrong and the url state doesn't update. 

i have a temp fix for that. a clean function that destroy the widget when the location of the page is changing.

Screenshot 2023-10-04 at 0 49 37

### PR DESCRIPTION
**Pull request by JoshuaRotimi**
This PR fixes the issue #64 with creating multiple instances of the `widget` component across renders
 - Adds a cleanup function to the `uploadWidget` component
 - States a warning in the console when the required env variables are missing to prevent confusion to users on why the Widget might be showing an endless spinner
 - Stores the widget instance in a ref 

These fixes were only applied on the React preset widget example. If all works well, they can be replicated for the Remix and Next.js examples. 

Fixes #64

### CODE REVIEW

**Comment by colbyfayock** [comment]
hey @JoshuaRotimi this definitely works as expected! i'm going to poke around early next week with this copy and pasted into a project with routing to see how the cleanup / variables are handled to make sure it's handling that use case before merging it in

**Comment by JoshuaRotimi** [comment]
Alright sounds great @colbyfayock. After your testing if all works well maybe we can implement the same for the other react centric examples before merging ?

**Comment by JoshuaRotimi** [comment]
Alright great! I will make the changes and update the PR

**Comment by colbyfayock** [comment]
hey @JoshuaRotimi id be curious if we have this issue with our Next Cloudinary library: https://github.com/colbyfayock/next-cloudinary/blob/main/next-cloudinary/src/components/CldUploadWidget/CldUploadWidget.tsx

we're not destroying the widget instance anywhere, however i dont see a similar issue navigating around page to page like we saw in this example

if you have any interest in playing with that and seeing if there are any issues, it could be another contribution opportunity

there's also a CldVideoPlayer component which I think has a bug open for a similar issue: https://github.com/colbyfayock/next-cloudinary/issues/279#issuecomment-1762134067

no pressure, but thought id share :)

**Comment by JoshuaRotimi** [comment]
Yes @colbyfayock , I just tested it out on the Next Cloudinary library and the same issue occurs. Perhaps you can create the issue and assign it to me and I will make the PR to fix it? As for the CldVideoPlayer component bug, I will check it out too.

**Comment by colbyfayock** [comment]
assigned you the ticket, thanks Josh! Let me konw when this current PR is ready to review again :)

**Comment by JoshuaRotimi** [comment]
Okay great! I will check it out @colbyfayock . Meanwhile This PR is ready to be reviewed again.

**Comment by colbyfayock** [comment]
looks like you're trying to commit an `.idea` directory, can you delete that then add `.idea` to the root `.gitignore`? you'll have to do it in separate commits, first deleting, then adding to `.gitignore`

File: examples/remix-upload-widget-preset/app/components/UploadWidget/index.jsx [hunk 1]
```diff
@@ -39,9 +43,17 @@ function UploadWidget({ children, onUpload }) {
         // widget without requiring an API Key or Secret. This however allows for
         // "unsigned" uploads which may allow for more usage than intended. Read more
         // about unsigned uploads at: https://cloudinary.com/documentation/upload_images#unsigned_upload
+
+        const cloudName = ENV.CLOUD_NAME;
```
Comment (colbyfayock) [inline examples/remix-upload-widget-preset/app/components/UploadWidget/index.jsx:47]: this should be `CLOUDINARY_CLOUD_NAME`

**Comment by colbyfayock** [comment]
i'm getting an error in `nextjs-upload-widget-preset` and `nextjs-upload-widget-signed` - did you happen to test these?

image

image

**Comment by JoshuaRotimi** [comment]
You were very correct! @colbyfayock . Seems like calling the `handleOnLoad` function in the useEffect caused the errors in the images above. I've removed them and also fixed the other issues mentioned. I believe you can test this PR again now. Also makes me wonder why that issue fixed the video rendering on the `CldVideoPlayer` component in the next-cloudinary example. I will take a closer look at that and let you know what I find!

**Comment by colbyfayock** [comment]
looks great @JoshuaRotimi thanks for the fixes!

**Comment by colbyfayock** [comment]
@all-contributors please add @JoshuaRotimi for code

**Comment by colbyfayock** [comment]
@JoshuaRotimi congrats on the merged PR! i can't remember if this is your first hacktoberfest contribution, but if it is your first, since it was merged during Hacktoberfest, you're eligible for some free swag. Please email [hacktoberfest@cloudinary.com](mailto:hacktoberfest@cloudinary.com) with your Github username and a link to this PR where we'll follow up with you there.

https://cloudinary.com/blog/hacktoberfest-celebrate-open-source-sdks

**Comment by JoshuaRotimi** [comment]
Thanks @colbyfayock This is my second PR so I've done that already. Always happy to contribute!

**Pull request merged**